QGL/drivers/APS3Pattern.py

```diff
@@ -0,0 +1,144 @@
+
+def write_sequence_file(awgData, fileName):
+    '''
+    Main function to pack channel sequences into an APS2 h5 file.
+    '''
+    # Convert QGL IR into a representation that is closer to the hardware.
+    awgData['ch1']['linkList'], wfLib = preprocess(
+        awgData['ch1']['linkList'], awgData['ch1']['wfLib'])
+
+    # compress marker data
+    for field in ['m1']:
+        if 'linkList' in awgData[field].keys():
+            PatternUtils.convert_lengths_to_samples(awgData[field]['linkList'],
+                                                    SAMPLING_RATE, 1,
+                                                    Compiler.Waveform)
+            compress_marker(awgData[field]['linkList'])
+        else:
+            awgData[field]['linkList'] = []
+
+    #Create the waveform vectors
+    wfInfo = []
+    wfInfo.append(create_wf_vector({key: wf.real
+                                    for key, wf in wfLib.items()}, awgData[
+                                        'ch1']['linkList']))
+    wfInfo.append(create_wf_vector({key: wf.imag
+                                    for key, wf in wfLib.items()}, awgData[
+                                        'ch1']['linkList']))
+
+    if SAVE_WF_OFFSETS:
+        #create a set of all waveform signatures in offset dictionaries
+        #we could have multiple offsets for the same pulse becuase it could
+        #be repeated in multiple cache lines
+        wf_sigs = set()
+        for offset_dict in wfInfo[0][1]:
+            wf_sigs |= set(offset_dict.keys())
+        #create dictionary linking entry labels (that's what we'll have later) with offsets
+        offsets = {}
+        for seq in awgData['ch1']['linkList']:
+            for entry in seq:
+                if len(wf_sigs) == 0:
+                    break
+                if isinstance(entry, Compiler.Waveform):
+                    sig = wf_sig(entry)
+                    if sig in wf_sigs:
+                        #store offsets and wavefor lib length
+                        #time ampltidue entries are clamped to ADDRESS_UNIT
+                        wf_length = ADDRESS_UNIT if entry.isTimeAmp else entry.length
+                        offsets[entry.label] = ([_[sig] for _ in wfInfo[0][1]],
+                                                wf_length)
+                        wf_sigs.discard(sig)
+
+            #break out of outer loop too
+            if len(wf_sigs) == 0:
+                break
+
+        #now pickle the label=>offsets
+        with open(os.path.splitext(fileName)[0] + ".offsets", "wb") as FID:
+            pickle.dump(offsets, FID)
+
+    # build instruction vector
+    seq_data = [awgData[s]['linkList']
+                for s in ['ch1', 'm1']]
+    instructions = create_instr_data(seq_data, wfInfo[0][1], wfInfo[0][2])
+
+    #Open the binary file
+    if os.path.isfile(fileName):
+        os.remove(fileName)
+
+    with open(fileName, 'wb') as FID:
+        FID.write(b'APS3')                     # target hardware
```

Comment:
a question of taste but I find using `struct.pack` nicer for this because you can be more explicit about widths and byte ordering

Reply:
Sure, I chose to use the numpy functions primarily because the data in question is already numpy floats and will be read back from the file into numpy floats, so rather than explicitly specifying the byte width / ordering and having to make sure it's commensurate with what numpy expects, using the numpy functions can handle this implicitly and leave less room for error.